### DISCUSSION TITLE
Presenting a Sequence of Images

### DISCUSSION THREAD

**Discussion by jswhitlo6** [jspsych/jsPsych#812] [Q&A] [answered]
I am attempted to program a simple study where an image is first presented, followed by another image 1/4 its size is presented in the middle of the previous image (i.e., First present image A, then Image B superimposed on the center of Image A). 

I used the following code based on the Lexical Decision task tutorial, where I altered it to present images instead of words. I can get it to present the first image, but then it crashes on me. 

```html
<!DOCTYPE html>
<html>
    <head>
        <title>Objects_Scenes</title>
        <script src="jspsych/jspsych.js"></script>
        <script src="jspsych/plugins/jspsych-html-keyboard-response.js"></script>
        <script src="jspsych/plugins/jspsych-image-keyboard-response.js"></script>
        <script src="jspsych/plugins/jspsych-instructions.js"></script>
        <link href="jspsych/css/jspsych.css" rel="stylesheet">
    </head>
    <body></body>
    <script>

        var timeline = [];

        var instructions = {
            type: 'instructions',
            pages: ['Welcome!','<p>You will be presented scenes, followed by obbjects superimposed on those scenes.</p><p>When you see each object, think of how well it goes with the scene.</p><p>Your memory for objects will be tested later</p><p>Click the next button to start.</p>'],
            show_clickable_nav: true
        };
        timeline.push(instructions);

        var fixation = {
            type: 'html-keyboard-response',
            stimulus: '<span style="font-size:40px;">+</span>',
            choices: jsPsych.NO_KEYS,
            trial_duration: 1000
        }
      
        var scene = {
            type: 'image-keyboard-response',
            stimulus: jsPsych.timelineVariable('scene'),
            stimulus_duration: 2000,
            choices: jsPsych.NO_KEYS,
            post_trial_gap: 0000
            };
        

        var object_scene = {
            type: 'image-keyboard-response',
            stimulus: jsPsych.timelineVariable('object'),
            stimulus_duration: 2000,
            choices: jsPsych.NO_KEYS,
            post_trial_gap: 1000
            };


        var trial_info = [
            {scene: 's001.jpg'},
            {scene: 's002.jpg'},
            {object: 'os001.jpg'},
            {object: 'os002.jpg'}
        ];

        var object_scene_procedure = {
            timeline: [fixation,scene,object_scene],
            timeline_variables: trial_info,
            randomize_order: false
        }
        timeline.push(object_scene_procedure);

        jsPsych.init({
            timeline: timeline,
            on_finish: function() {
                jsPsych.data.displayData();
            }
        })

</script>

</html>
```

**Comment by conbainbridge** [comment, marked as answer]
Because there are no keys being used to continue the trial, you additionally will have to add a "trial_duration" property in addition to the stimulus_duration.

**Reply by jswhitlo6** [reply]
What plugin would you suggest using if I want to present images without requiring the participant to make any responses on the keyboard?

**Reply by conbainbridge** [reply]
I think the way you have it setup works well for this! The stimulus_duration just doesn't impact the trial_duration so the two have to both be defined in this case.

**Reply by jswhitlo6** [reply]
Looks like the stimulus_duration did the trick! Thank you so much!

Another question. I was able to successfully present all images that are in my trial_info table. There are a number of things I noticed that don't seem to be following the code. 

So on each trial I have two variables for presenting the images. One for image A, next for image A+B.  Instead of treating it as one trial, it's breaking it up into two trials (the fixation point at the beginning of the trial is presented between the two stimuli presentations). 
The timing is also affected. When I program Image A to be present for 2000 ms, and Image B for 4000 ms, it sets the timing for both stimuli presentations to be 4000 ms. 

These are probably naive questions. I'm new to javascript so this is probably something really small. Provided below is the code. Basically what I need it to do is go from Image A to Image B uninterrupted in one trial, with no ISI between the two image presentations. 

```html
<!DOCTYPE html>
<html>
    <head>
        <title>Objects_Scenes</title>
        <script src="jspsych/jspsych.js"></script>
        <script src="jspsych/plugins/jspsych-html-keyboard-response.js"></script>
        <script src="jspsych/plugins/jspsych-image-keyboard-response.js"></script>
        <script src="jspsych/plugins/jspsych-instructions.js"></script>
        <link href="jspsych/css/jspsych.css" rel="stylesheet">
    </head>
    <body></body>
    <script>

        var timeline = [];

        var instructions = {
            type: 'instructions',
            pages: ['Welcome!','<p>You will be presented scenes, followed by obbjects superimposed on those scenes.</p><p>When you see each object, think of how well it goes with the scene.</p><p>Your memory for objects will be tested later</p><p>Click the next button to start.</p>'],
            show_clickable_nav: true
        };
        timeline.push(instructions);

        var fixation = {
            type: 'html-keyboard-response',
            stimulus: '<span style="font-size:40px;">+</span>',
            choices: jsPsych.NO_KEYS,
            trial_duration: 1000
        }
      
        var scene = {
            type: 'image-keyboard-response',
            stimulus: jsPsych.timelineVariable('stimuli'),
            stimulus_duration: 2000,
            choices: jsPsych.NO_KEYS,
            trial_duration: 2000
            };
        

        var object_scene = {
            type: 'image-keyboard-response',
            stimulus: jsPsych.timelineVariable('stimuli'),
            stimulus_duration: 4000,
            choices: jsPsych.NO_KEYS,
            trial_duration: 4000
            };


        var trial_info = [
            {stimuli: 's001.jpg'},
            {stimuli: 'os001.jpg'},
            {stimuli: 's002.jpg'},
            {stimuli: 'os002.jpg'}
        ];

        var object_scene_procedure = {
            timeline: [fixation,scene,object_scene],
            timeline_variables: trial_info,
            randomize_order: false
        }
        timeline.push(object_scene_procedure);

        jsPsych.init({
            timeline: timeline,
            on_finish: function() {
                jsPsych.data.displayData();
            }
        })

</script>

</html>
```

**Reply by jodeleeuw** [reply]
Hey @jswhitlo6 just FYI if you add three back ticks before your code and after your code it will format it nicely:

\`\`\`html
my html code here
\`\`\`

**Reply by conbainbridge** [reply]
I believe what is happening is that the use of two trials with the same "jsPsych.timelineVariable('stimuli')" is confusing the timeline_variables behind-the-scenes.

Depending on how you want to structure your study the best solution may vary, but one thing you could do is have multiple properties in your trial_info, such as "[{stimuliS: 's001.jpg', stimuliOS: 'os001.jpg'}, {stimuliS: 's002.jpg', stimuliOS: 'os002.jpg'}]", then have the trials use jsPsych.timelineVariable('stimuliS') and jsPsych.timelineVariable('stimuliOS').

**Reply by jswhitlo6** [reply]
The structure you suggested was one that I employed initially, but it had its own set of problems, including breaking it up into two trials (similar to what I'm experiencing now). My impression is that the software isn't designed to present two separate images within the same trial, since no matter how I structure the code, it always breaks it up into two separate trials.

**Reply by jodeleeuw** [reply]
You can use the `html-keyboard-response` plugin to show multiple images. The `image-keyboard-response` is restricted to one at a time, since the `html-keyboard-response` is much more flexible. 

See this discussion for an example: https://github.com/jspsych/jsPsych/discussions/802#discussioncomment-32656

**Reply by jswhitlo6** [reply]
Thank you for this. I see that this is for presenting two images simultaneously. Is there a way to present one image directly on top of another, but have one image be 1/4 the size of the other? As in Image A would be the background scene, and Image B would be superimposed on the center of Image A, but a 1/4 the size of Image A. 

I also need a way to present two images in a row with no ISI. Is there a way to present one image followed immediately by another with no ISI? That is, I want to present Image A, followed immediately by Image B. I thought it would be sufficient to simply code two variables for one trial, with one presenting one image and the next variable presenting another image. However, the program always breaks it up into two separate trials.

**Reply by jodeleeuw** [reply]
> Is there a way to present one image directly on top of another, but have one image be 1/4 the size of the other? As in Image A would be the background scene, and Image B would be superimposed on the center of Image A, but a 1/4 the size of Image A.

Yes, you can do this with HTML and CSS. It'll take a little bit of positioning with CSS. This might be a reasonable starting point:

```js
var trial = {
	type: 'html-keyboard-response',
	stimulus: '<div style="width:600px; height: 600px; position: relative;">'+
		'<img src="a.png" style="width:600px; height:600px; position: absolute; top: 0px; left: 0px;"></img>'+
		'<img src="b.png" style="width:150px; height: 150px; position: absolute: top: 225px; left: 225px;"></img>'+
		'</div>'
}
```

**Reply by jodeleeuw** [reply]
> Is there a way to present one image followed immediately by another with no ISI? That is, I want to present Image A, followed immediately by Image B. I thought it would be sufficient to simply code two variables for one trial, with one presenting one image and the next variable presenting another image. However, the program always breaks it up into two separate trials.

By default there is no ISI between trials, so simply having two consecutive trials should result in an ISI of 0, as long as the images are preloaded. Here's [documentation on preloading](https://www.jspsych.org/overview/media-preloading/).

**Reply by jswhitlo6** [reply]
I was able to successfully implement the code that you provided, thank you so much Josh! 

The preloading resource looks helpful as well. 

Ideally, I would be able to implement the code that you provided to reference a table of stimuli, so that I can put a loop for it to run through an entire stimuli set. I assume this could be coordinated with preloading my stimuli as well.

**Reply by jodeleeuw** [reply]
> Ideally, I would be able to implement the code that you provided to reference a table of stimuli, so that I can put a loop for it to run through an entire stimuli set. I assume this could be coordinated with preloading my stimuli as well.

You can do this with [timeline variables](https://www.jspsych.org/overview/timeline/#timeline-variables). See the second example under that section for a pretty close analog.
